interface to derivatives of 2-point loop functions

wkotlarski opened the pull request:
Things to note:

- `Softsusy` and `fflite` do not provide derivatives. We create dummy functions that return quiet NaN as we already did in the case of functions computing all PV functions with given number of legs at once.

wkotlarski commented:
@Expander, there's a design decision to be made. `LoopTools` can return all `B` functions and their derivatives at once. `Collier` can't. Should we try to create a function returning all loop function and their derivatives? In other words, should our interface match `LoopTools` or `Collier`?

Expander commented:
> @Expander, there's a design decision to be made. `LoopTools` can return all `B` functions and their derivatives at once. `Collier` can't. Should we try to create a function returning all loop function and their derivatives? In other words, should our interface match `LoopTools` or `Collier`?

That's a good question. First of all I think the basic interface should be as flexible as possible, such that we can (in principle) also use other libraries. So, this would be a point for having separate functions for each derivative.

Another perspective is performance: If one needs (for some application) only a specific derivative (as for example in FlexibleEFTHiggs, where we need only the derivative w.r.t. p^2), it would in general be best to only compute this derivative, and not all derivatives.

However, I can imagine that there is a reason for the additional `LoopTools` function that returns all derivatives: Maybe it reuses sub-expressions, which can speed-up the calculation if one needs all derivatives (similarly to the `sincos` function provided by some libraries). So, what do you think of providing such an interface function in addition? It would be trivial to implement for `LoopTools` and `COLLIER` and could provide a benefit if we need all derivatives at once?

wkotlarski commented:
We will have separate functions. In fact that is already implemented and tested. The point is that we had two types of calls to loop libraries

- giving a single loop function, e.g. `B0`
- giving all loop functions as a vector, `B0`, B1` etc

The logic as explained in `LoopTools` manual is that once you compute for example `B0` for certain kinematics you more or less also have all the other ones. The question is about this function returning a vector. Sorry, I was not very clear. Should we have two functions: one returning a vector of B's and one returning a vector of derivatives. `LoopTools` returns one huge vector of loop functions *and* their derivatives. In Collier we'd have to make two calls and combine results. There are pros and cons of both approaches (I cannot say which one would be faster)

- In `LoopTools` you always get all of them. So it's free
- Otoh, you usually don't need both loop functions and derivatives at the same time. So in Collier's case that would be a slowdown.

I'm leaning into a solution where we have two functions: one returning a vector of B's and one returning a vector of their derivatives.

Ps. I'm not sure what you mean by other derivatives. We want only derivatives w.r.t. $p^2$.

Expander commented:
Sorry, I misunderstood.

> I'm leaning toward a solution where we have two functions: one returning a vector of B's and one returning a vector of their derivatives.

I'd also vote for such a "split" approach. We never had many applications where one needed the loop functions and their derivatives at the same time.

> Ps. I'm not sure what you mean by other derivatives. We want only derivatives w.r.t. p2.

Ah, ok. Well, in principle for the function $B(s,x,y)$ we could have derivatives w.r.t. all three parameters. So far, we only needed derivatives w.r.t. $p^2$, as you said.

wkotlarski commented:
I'm mostly done. Question about your recent changes. You've introduced a new function to `mass_eigenstates.cpp`
```cpp
/// derivative of B0(p^2, ma^2, mb^2) w.r.t. p^2
double CLASSNAME::DB0(double p2, double m2a, double m2b) const noexcept
{
   return softsusy::db0(p2, m2a, m2b);
}
```
I've replaced it with
```cpp
double CLASSNAME::DB0(double p2, double m2a, double m2b) const noexcept
{
   const double scl2 = Sqr(get_scale());
   return Loop_library::get().DB0(p2, m2a, m2b, scl2).real();
}
```
For this to work with `softsusy` as loop library I've introduced
```cpp
std::complex<double> Softsusy::DB0(DB_ARGS) noexcept
{
   double p = std::sqrt(p10_in.real());
   double m1 = std::sqrt(m02_in.real());
   double m2 = std::sqrt(m12_in.real());

   return {softsusy::db0(p, m1, m2), 0.0};
}
```
This however means that `DB0` behaves differently if you use `LoopTools` or `Collier` and `softsusy` (the later one only gives derivative at $p^2=0$). Are you ok with that?

wkotlarski commented:
One more things. Why are loop function in `mass_eigenstates` defined as double?

Expander commented on templates/mass_eigenstates.cpp.in:
```diff
@@ -462,18 +462,36 @@ double CLASSNAME::G0(double p, double m1, double m2) const noexcept
 
-/// derivative of B0(p^2, ma^2, mb^2) w.r.t. p^2
-double CLASSNAME::DB0(double p2, double m2a, double m2b) const noexcept
+double CLASSNAME::DH0(double p, double m1, double m2) const noexcept
```
Please add a brief comment her that it is the derivative of H0 w.r.t. p^2 and the arguments are of dimension mass squared.

wkotlarski replied:
Done in 46963d81122636e42cb7c82d7f26193087090e7d.

Expander commented on templates/mass_eigenstates.cpp.in:
```diff
@@ -462,18 +462,36 @@ double CLASSNAME::G0(double p, double m1, double m2) const noexcept
+   return 4.0*Loop_library::get().DB00(p, m1, m2, scl2).real() + CLASSNAME::DG0(p, m1, m2);
+}
+
+double CLASSNAME::DF0(double p, double m1, double m2) const noexcept
```
Same here, please add a brief comment.

wkotlarski replied:
Done in 46963d81122636e42cb7c82d7f26193087090e7d.

Expander commented:
One general comment, which is slightly related: We define the symbols `A_N`, `B_N` etc. in a a loop library header file. Since these preprocessor symbols are not protected in any way (no scope, no type), I'd suggest to rename them as `FS_A_N`, i.e. add a prefix like `FS_`, or even `FLEXIBLESUSY_` or so. (This should probably be done in a separate PR.)

uukhas reviewed (changes requested):
Thank you very much for adding these missing functions. I have only several small comments/questions:

- [ ]  Sometimes you use B_ARGS, sometimes DB_ARGS for derivatives. Though the second macro is defined as the first one, I believe, that it would be better to use DB everywhere. Also, at one point there was a usage of C and B things, which looks suspicious. (I made comments in the code, so that, hopefully, you can see them immediately.)
- [ ] I remember that in the code that T. did he needed the derivatives of B functions and at that point it was faster to implement them independently from my interface, did you try to update all of the derivatives in FS to the new format?
* Thank you for changing my confusing naming in the documentation!
- [ ] Oh, finally, do you think it would be better to update our wiki so that everyone knows how to add new functions? Or even documentation. As now you remember and know how to do it, it can be useful in future, what do you think?

So, currently, I choose "Request changes" option as it feels more natural now :)

wkotlarski commented on src/loop_libraries/library_looptools.hpp:
```diff
@@ -38,10 +38,12 @@ class Looptools : public Loop_library_interface
    Looptools();
    BOOST_PP_SEQ_FOR_EACH(REDEFINE, (A_ARGS), A_SEQ)
    BOOST_PP_SEQ_FOR_EACH(REDEFINE, (B_ARGS), B_SEQ)
+   BOOST_PP_SEQ_FOR_EACH(REDEFINE, (B_ARGS), DB_SEQ)
```
Done in c7597f9458024577e3e17b64164bf89e233a5c3a.

wkotlarski commented on src/loop_libraries/library_softsusy.hpp:
```diff
@@ -33,10 +33,12 @@ class Softsusy : public Loop_library_interface
 public:
    BOOST_PP_SEQ_FOR_EACH(REDEFINE, (A_ARGS), A_SEQ)
    BOOST_PP_SEQ_FOR_EACH(REDEFINE, (B_ARGS), B_SEQ)
+   BOOST_PP_SEQ_FOR_EACH(REDEFINE, (B_ARGS), DB_SEQ)
```
Done in 449c767612f386c5e65829155afd4cf1eec0d040.

wkotlarski commented on src/loop_libraries/library_fflite.hpp:
```diff
@@ -34,6 +34,7 @@ class Fflite : public Loop_library_interface
    Fflite();
    BOOST_PP_SEQ_FOR_EACH(REDEFINE, (A_ARGS), A_SEQ)
    BOOST_PP_SEQ_FOR_EACH(REDEFINE, (B_ARGS), B_SEQ)
+   BOOST_PP_SEQ_FOR_EACH(REDEFINE, (B_ARGS), DB_SEQ)
```
Done in 449c767612f386c5e65829155afd4cf1eec0d040.

wkotlarski commented on src/loop_libraries/library_fflite.cpp:
```diff
@@ -83,6 +83,7 @@ std::complex<double> Fflite::B00(B_ARGS) noexcept
    return cb2i[1];
 }
 
+BOOST_PP_SEQ_FOR_EACH(UNDEFINED, (C_ARGS), DB_SEQ)
```
Done in c05b954547e71b6050529a1209f37c1621156af7.

wkotlarski commented on src/loop_libraries/library_collier.hpp:
```diff
@@ -39,10 +39,12 @@ class Collier : public Loop_library_interface
    Collier() : current_mu2_uv(1.0) { initialize(); }
    BOOST_PP_SEQ_FOR_EACH(REDEFINE, (A_ARGS), A_SEQ)
    BOOST_PP_SEQ_FOR_EACH(REDEFINE, (B_ARGS), B_SEQ)
+   BOOST_PP_SEQ_FOR_EACH(REDEFINE, (B_ARGS), DB_SEQ)
```
Done in 449c767612f386c5e65829155afd4cf1eec0d040.

wkotlarski commented on src/loop_libraries/library_collier.cpp:
```diff
@@ -129,6 +132,7 @@ std::complex<double> Collier::A0(A_ARGS) noexcept
 }
 
 BOOST_PP_SEQ_FOR_EACH(COLLIER_B, (B_ARGS), B_SEQ)
+BOOST_PP_SEQ_FOR_EACH(COLLIER_B, (DB_ARGS), DB_SEQ)
```
This one is fine. `COLLIER_B` is the name of the whole macro, which is the same for `B` and `DB` functions.

wkotlarski commented:
> I remember that in the code that T. did he needed the derivatives of B functions and at that point it was faster to implement them independently from my interface, did you try to update all of the derivatives in FS to the new format?

Yes. I've replaced the call to `softsusy::db0` in `mass_eigenstates.cpp` with the call to your loop library interface.
